### PR TITLE
WIP: Proposal to implement #1615 by abusing the type-checker

### PR DESCRIPTION
**Pull request by AdamSaleh**
I looked into a possibility of implementing the code-completion for record's labels and I woul like to know your opinion.

* I changed the parser to allow for empty labels, and now when I have annotated record, It will throw a type error instead of parse-error, when I start writing empty label. This is a terrible idea, because suddenly, you can have valid records `{ x = 1,}` of type `{``:Bool, x: Natural}`, but I didn't want to attempt to fix that glaring hole before I ask if the general attempt holds at least some water.
* now I can use the type-error in the completion handler when in the incomplete record.

What do you think?

### CODE REVIEW

**Comment by Gabriel439** [comment]
@AdamSaleh: Using the type error to infer record completions is really clever.

I think the main change in tack I would suggest is to implement the parsing hack on the `dhall-lsp-server` side instead of within the `dhall` code.  Specifically, what the language server can do is:

* First attempt to parse and type-check the code as-is

* If parsing fails, delete code up to and including the leading comma before the cursor and then attempt to parse and type-check the code again

* If the second attempt gives an annotation mismatch error for a record, then convert that into completions for the user.  Otherwise, reuse the parsing failure from the first attempt

**Comment by AdamSaleh** [comment]
If anybody is wondering, I would still like to try to complete this, so I am giving myself the deadline of 14. of June, since that was supposed to have been Zurihac, where I wuld have definitely found some time to work on this :-)

**Comment by Gabriel439** [comment]
@AdamSaleh: No rush!  🙂

**Comment by PierreR** [comment]
@AdamSaleh If you can improve `dhall-lsp-server` that would be fantastic. Currently I can't use it probably because of https://github.com/dhall-lang/dhall-haskell/issues/1804. It is unbearably slow (minutes to do anything useful).

That said the graph from my import is currently huge. It contains dhall-openshift (which in turn contains dhall-kubernetes), the whole Prelude and some custom object. I probably need to find a way to cut the graph somehow. Or maybe `dhall-lsp-server` could be more lazy ?

**Comment by PierreR** [comment]
I have created https://github.com/dhall-lang/vscode-dhall-lsp-server/issues/20 in case it helps to track this down (or improve the situation somehow)

**Comment by AdamSaleh** [comment]
Another take. Only using the type error information, will complete `, fieldname`, including the leading comma.

**Comment by sjakobi** [comment]
@AdamSaleh Apologies, I haven't followed the discussion on the ZuriHac Discord in detail.

Could you summarize the current status?

From what I can see now, the main thing that would be good to have would be some documentation (also in the `dhall-lsp-server` readme) and possibly a bit more test coverage.

**Comment by AdamSaleh** [comment]
It looks like w.r.t. using the type-error I was trying to be too clever and the usability was bad (even though I found the thing somewhat useful).

I am fine with adding more tests coverage, but first I want my record completion, dammit! :D

My next stab probably will be abandoning the idea with type-check-errors, and focusing on schemas, because searching for the schema should be doable as well.

**Comment by AdamSaleh** [comment]
Ok, I am changing the thing to no longer use type-errors, and just using the `Schema::` and that probably warrants new PR with better description :-)

**Comment by sjakobi** [comment]
I was wondering whether the standard change to allow trailing commas (https://github.com/dhall-lang/dhall-lang/pull/956) might be useful in this context, but I'm also way out-of-date with your implementation strategy. ;)
